tests/unit/bootstrap_record_test.cc

```diff
@@ -0,0 +1,269 @@
+        std::memcpy(buffer, buf.data() + pos, len);
+        return make_ready_future<size_t>(len);
+    }
+
+    future<> flush() override {
+        return make_ready_future<>();
+    }
+
+    future<> close() override {
+        return make_ready_future<>();
+    }
+};
+
+future<block_device> prepare_device() {
+    return async([&] {
+        sstring device_path = [&]() -> sstring {
+            temporary_file tmp_device_file("/tmp/bootstrap_record_test_file");
+            return tmp_device_file.path();
+        }();
+
+        file f = open_file_dma(device_path, open_flags::rw | open_flags::create).get0();
+        f.truncate(device_size).get();
+        f.close().get();
+
+        return open_block_device(device_path).get0();
+    });
+}
+
+}
+
+BOOST_TEST_DONT_PRINT_LOG_VALUE(bootstrap_record)
+SEASTAR_TEST_CASE(test_valid_read_write) {
+    return prepare_device().then([] (block_device dev) {
+        return do_with(std::move(dev), [] (block_device& dev) {
+            return async([&] {
+                bootstrap_record write_record(1, 1024, 1024, 1, 3, {{6, {6, 9}}, {9, {9, 12}}, {12, {12, 15}}});
+                // basic test
+                write_record.write_to_disk(dev).get();
+                auto read_record = bootstrap_record::read_from_disk(dev).get0();
+                BOOST_REQUIRE_EQUAL(write_record, read_record);
+
+                // max number of shards
+                write_record.shards_nb = bootstrap_record::max_shards_nb;
+                write_record.shards_info.resize(write_record.shards_nb, {1, {2, 3}});
+                write_record.write_to_disk(dev).get();
+                read_record = bootstrap_record::read_from_disk(dev).get0();
+                BOOST_REQUIRE_EQUAL(write_record, read_record);
+
+                // 0 shards
+                write_record.shards_nb = 0;
+                write_record.shards_info.clear();
+                write_record.write_to_disk(dev).get();
+                read_record = bootstrap_record::read_from_disk(dev).get0();
+                BOOST_REQUIRE_EQUAL(write_record, read_record);
+            }).finally([&] {
+                return dev.close();
+            });
+        });
+    });
+}
+
+namespace {
+
+void repair_crc32(shared_ptr<mock_block_device_impl> dev_impl) {
+    auto crc32 = [](const uint8_t* buff, size_t len) {
+        boost::crc_32_type result;
+        result.process_bytes(buff, len);
+        return result.checksum();
+    };
+    mock_block_device_impl::buf_type& buff = dev_impl.get()->buf;
+    size_t crc_pos = offsetof(bootstrap_record_disk, crc);
+    uint32_t crc_new = crc32(buff.data(), crc_pos);
+    std::memcpy(buff.data() + crc_pos, &crc_new, sizeof(crc_new));
+}
+
+void change_byte_at_offset(shared_ptr<mock_block_device_impl> dev_impl, size_t offset) {
+    dev_impl.get()->buf[offset]++;
+}
+
+template<typename T>
+void change_bytes_at_offset_to(shared_ptr<mock_block_device_impl> dev_impl, size_t offset, T value) {
```

Comment:
```suggestion
inline void change_bytes_at_offset_to(shared_ptr<mock_block_device_impl> dev_impl, size_t offset, T value) noexcept {
```

Reply:
Maybe rename to something like `place_at_offset` because `bytes` suggest that value is `bytes` in some sense (a typical situation would be decltype(value) == const uint8_t*)